MINOR: Automatic async refreshing of connection state did not send internal events

rhauch opened the pull request:
## Summary of Changes

We recently added an asynchronous refreshing of connection states. Any changes in the connection state (e.g., from `ATTEMPTING` to `CONNECTED`, we would not fire the local events indicating the change in connectivity.

This fixes that and encapsulates the change in state to ensure the `StateChangedListener` on each `ConnectionState` object is called when the status changes.
* A CONNECTED internal event is fired when a connection’s state was previously not connected (see below) but becomes connected, or when it was previously connected and the state has changed in some way.
* A DISCONNECTED internal event is fired when a connection’s state was previously connected (see below) but becomes disconnected, or when it was not previously connected and the state has changed in some way.

This logic relies upon the `ConnectionState.isConnected()` method to determine whether the connection is connected or not connected.


## Pull request checklist

Please check if your PR fulfills the following (if applicable):

- Tests:
    - [ ] Added new
    - [ ] Updated existing
    - [ ] Deleted existing
- [ ] Have you validated this change locally against a running instance of the Quarkus dev server?
    ```shell
    make quarkus-dev
    ```
- [ ] Have you validated this change against a locally running native executable?
    ```shell
    make mvn-package-native && ./target/ide-sidecar-*-runner
    ```

flippingbits commented:
Thanks, @rhauch! This looks great.

> We recently added an asynchronous refreshing of connection states. Any changes in the connection state (e.g., from ATTEMPTING to CONNECTED, we would not fire the local events indicating the change in connectivity.

When we migrated to asynchronously refreshing the connection statuses, we did not change how events indicating updates in the connection states are emitted. CCloud connections emit local events for a few cases (see [here](https://github.com/confluentinc/ide-sidecar/blob/main/src/main/java/io/confluent/idesidecar/restapi/connections/CCloudConnectionState.java#L122) and [here](https://github.com/confluentinc/ide-sidecar/blob/main/src/main/java/io/confluent/idesidecar/restapi/connections/CCloudConnectionState.java#L138)), while other connection types don't emit events at all.

I like your approach because it emits these local events in a more generic fashion for all connection types. Also, it emits events only if the status has really changed. Before this PR, the sidecar would emit events whenever the status of a CCloud connection is refreshed.

Do you mind removing the emitting of events from the `CCloudConnectionState` so that we don't produce duplicate events?

flippingbits commented on src/main/java/io/confluent/idesidecar/restapi/connections/ConnectionState.java:
```diff
@@ -100,8 +100,38 @@ protected ConnectionStatus getInitialStatus() {
+    // update the cached status
+    this.cachedStatus.set(updated);
+
+    // Figure out which event to send
+    if (updated.isConnected()) {
+      // Now connected
+      if (!original.isConnected()) {
+        // but was not connected before so we connected
+        listener.connected(this);
+      } else if (!original.equals(updated)) {
+        // and was connected before, but the status changed
+        listener.connected(this);
+      }
+    } else {
+      // Now not connected
+      if (original.isConnected()) {
+        // but was connected before so we disconnected
+        listener.disconnected(this);
+      } else if (!original.equals(updated)) {
+        // and was not connected before, but the status changed
+        listener.disconnected(this);
+      }
+    }
```
Is it safe to simplify the code as follows?
```suggestion
    // Send an event if the status has changed
    if (!updated.equals(original)) {
      if (updated.isConnected()) {
        listener.connected(this);
      } else {
        listener.disconnected(this);
      }
    }
```

rhauch commented:
Thanks for the feedback and suggestions, @flippingbits. I've incorporated your suggestions, and added a few more unit tests:
* The `ConnectionStateManager` now fires a connected event if the new connection is already connected (e.g., local connections).
* Make the `ConnectionState.listener` private so as to prevent subclasses for directly calling it, as the listener should only be called when lifecycle events occur (in `ConnectionStateManager`) or when the status changes (via `ConnectionState.refreshStatus()`.
* Removed from `CCloudConnectionState` and `LocalConnectionState` the calls to the listener, as they are no longer needed due to the above changes.
* Added a few new unit tests to verify how `ConnectionState.refreshStatus()` should or should not call the listener.

flippingbits reviewed (approved):
Thanks for the update, @rhauch. LGTM! 🎉
